Convert path extension to pybind11

QuLogic opened the pull request:
## PR summary

Another part of #23846, but waiting for some other refactoring PRs to be ready.

## PR checklist

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [n/a] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [n/a] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

tacaswell commented:
Given the scope of these changes, what is the best way to review these?

QuLogic commented:
I think commit-by-commit is not bad, or maybe I should split out the first commit from the remaining ones, as those are all fairly small.

But I also think we maybe should wait for #26050, which I want to get back to reviewing, and I don't want to cause some undue rebasing for the contributor.

ianthomas23 commented:
I should have some time available in a week or two to help review pybind11/C++ changes, if help is required?

oscargus commented:
Maybe we can consider #25645 as well (after #26050) before this? Although I can probably recreate it (rather than rebasing) if there are any issues.

ianthomas23 commented:
@QuLogic Both this and the other pybind11 PR (#27011) are labelled as "waiting for other PR". Can you clarify what needs to be done before either of them become reviewable?

QuLogic commented:
@ianthomas23 I was waiting for #26050 to make it easier for the new contributor, but as @story645 says she will take care of it, I've rebased this PR and fixed the conflicts.

ianthomas23 commented:
@QuLogic Reviewing this is on my todo list, hopefully within the next week. Feel free to ping me if I forget.

ianthomas23 reviewed (approved):
Just a few comments and style issues.

QuLogic commented on src/py_converters_11.h:
```diff
@@ -8,8 +8,145 @@
+                return true;
+            }
+
+            auto rect_arr = py::array_t<double>::ensure(src);
```
I don't think so; we access the array using pybind11 wrappers only, and nothing like `memcpy` or passing it to something that expects an array.

QuLogic commented on src/_path_wrapper.cpp:
```diff
@@ -1,138 +1,91 @@
-        dims[1] = 2;
-
-        dims[0] = (npy_intp)poly.size();
+        py::ssize_t dims[] = { (py::ssize_t)poly.size(), 2 };
```
Oh, I thought I only saw C-style casts for dimensions in other places, but those must've been replaced by `static_cast` at some point, so I'll change here as well.

QuLogic commented on src/_path_wrapper.cpp:
```diff
@@ -148,570 +101,302 @@ static PyObject *Py_update_path_extents(PyObject *self, PyObject *args)
 {
     bool result;
 
     PyArrayObject *array = (PyArrayObject *)PyArray_CheckFromAny(
```
I thought I had tried this, but I can't find a stash for it. There are still a few other `numpy::array_view` in this file though, so I think removing just this one won't mean removing `import_array`?

For those other calls, we still need to get Agg closer to pybind11, and then we can do the NumPy API removal, so I think I'll wait on this one.

ianthomas23 replied:
Sure.

QuLogic commented on src/_path_wrapper.cpp:
```diff
@@ -1,138 +1,91 @@
+    auto result = py::list(polygons.size());
 
     for (size_t i = 0; i < polygons.size(); ++i) {
         Polygon poly = polygons[i];
```
Fixed.

ianthomas23 commented:
This looks good to me, just needs a second review now.

QuLogic commented:
I've found that [pybind11 can do the copy to `py::array` for you](https://github.com/pybind/pybind11/issues/1042#issuecomment-325941022), so I've made that small change to remove the manual `memcpy`. It did need a cast for `Polygon`, as the otherwise it tries to make an array containing `XY` records (which aren't exposed).

QuLogic commented on src/_path_wrapper.cpp:
```diff
@@ -1,138 +1,111 @@
+#include <string>
+#include <vector>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
```
This should be moved first so that Cygwin doesn't break, cf #27821.

tacaswell reviewed (approved):
The tests pass, but we should merge the cgywin PR first.

QuLogic commented:
OK, looks like Cygwin is also happy, so I'll merge.
